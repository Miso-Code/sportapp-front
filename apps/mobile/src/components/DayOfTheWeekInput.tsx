import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { View, StyleSheet } from 'react-native'
import {
	Text,
	useTheme,
	Avatar,
	MD3Theme,
	TouchableRipple
} from 'react-native-paper'

const DAYS = [
	{ key: 'sunday', label: 'SUNDAY' },
	{ key: 'monday', label: 'MONDAY' },
	{ key: 'tuesday', label: 'TUESDAY' },
	{ key: 'wednesday', label: 'WEDNESDAY' },
	{ key: 'thursday', label: 'THURSDAY' },
	{ key: 'friday', label: 'FRIDAY' },
	{ key: 'saturday', label: 'SATURDAY' }
]

const localesStartingMonday = [
	'af-ZA', // Afrikaans (South Africa)
	'am-ET', // Amharic (Ethiopia)
	'ar-DZ', // Arabic (Algeria)
	'ar-LY', // Arabic (Libya)
	'ar-MA', // Arabic (Morocco)
	'az-Cyrl-AZ', // Azerbaijani (Cyrillic, Azerbaijan)
	'az-Latn-AZ', // Azerbaijani (Latin, Azerbaijan)
	'be-BY', // Belarusian (Belarus)
	'bg-BG', // Bulgarian (Bulgaria)
	'bn-BD', // Bengali (Bangladesh)
	'bs-Cyrl-BA', // Bosnian (Cyrillic, Bosnia and Herzegovina)
	'bs-Latn-BA', // Bosnian (Latin, Bosnia and Herzegovina)
	'ca-ES', // Catalan (Spain)
	'cs-CZ', // Czech (Czech Republic)
	'cy-GB', // Welsh (United Kingdom)
	'da-DK', // Danish (Denmark)
	'de-AT', // German (Austria)
	'de-CH', // German (Switzerland)
	'de-DE', // German (Germany)
	'de-LI', // German (Liechtenstein)
	'de-LU', // German (Luxembourg)
	'el-CY', // Greek (Cyprus)
	'el-GR', // Greek (Greece)
	'en-AU', // English (Australia)
	'en-GB', // English (United Kingdom)
	'en-IE', // English (Ireland)
	'en-NZ', // English (New Zealand)
	'es', // Spanish
	'es-CO', // Spanish (Colombia)
	'es-ES', // Spanish (Spain)
	'et-EE', // Estonian (Estonia)
	'eu-ES', // Basque (Spain)
	'fi-FI', // Finnish (Finland)
	'fo-FO', // Faroese (Faroe Islands)
	'fr-BE', // French (Belgium)
	'fr-CH', // French (Switzerland)
	'fr-FR', // French (France)
	'fr-LU', // French (Luxembourg)
	'ga-IE', // Irish (Ireland)
	'gl-ES', // Galician (Spain)
	'hr-HR', // Croatian (Croatia)
	'hu-HU', // Hungarian (Hungary)
	'hy-AM', // Armenian (Armenia)
	'is-IS', // Icelandic (Iceland)
	'it-CH', // Italian (Switzerland)
	'it-IT', // Italian (Italy)
	'ka-GE', // Georgian (Georgia)
	'kk-KZ', // Kazakh (Kazakhstan)
	'kl-GL', // Kalaallisut (Greenland)
	'lt-LT', // Lithuanian (Lithuania)
	'lv-LV', // Latvian (Latvia)
	'mk-MK', // Macedonian (North Macedonia)
	'mn-MN', // Mongolian (Mongolia)
	'nb-NO', // Norwegian BokmÃ¥l (Norway)
	'nl-BE', // Dutch (Belgium)
	'nl-NL', // Dutch (Netherlands)
	'nn-NO', // Norwegian Nynorsk (Norway)
	'pl-PL', // Polish (Poland)
	'pt-PT', // Portuguese (Portugal)
	'ro-RO', // Romanian (Romania)
	'ru-RU', // Russian (Russia)
	'sk-SK', // Slovak (Slovakia)
	'sl-SI', // Slovenian (Slovenia)
	'sq-AL', // Albanian (Albania)
	'sr-Cyrl-RS', // Serbian (Cyrillic, Serbia)
	'sr-Latn-RS', // Serbian (Latin, Serbia)
	'sv-FI', // Swedish (Finland)
	'sv-SE', // Swedish (Sweden)
	'tr-TR', // Turkish (Turkey)
	'uk-UA' // Ukrainian (Ukraine)
]

const DayOfTheWeekInput = ({ value, onChange, disabled = false, label }) => {
	const theme = useTheme()
	const { t, i18n } = useTranslation()
	const styles = createStyles(theme)

	const [selectedDays, setSelectedDays] = useState(value)

	if (
		localesStartingMonday.includes(i18n.language) &&
		DAYS[0].key !== 'monday'
	) {
		DAYS.push(DAYS.shift() as (typeof DAYS)[number]) // Move Sunday to the end
	}

	const handleDayPress = (day: string) => {
		if (selectedDays.includes(day)) {
			setSelectedDays(selectedDays.filter((d) => d !== day))
		} else {
			setSelectedDays([...selectedDays, day])
		}
	}

	useEffect(() => {
		if (onChange) onChange(selectedDays)
	}, [selectedDays, onChange])

	return (
		<>
			<Text variant='bodyMedium' style={disabled && styles.disabled}>
				{label}
			</Text>
			<View style={styles.container}>
				{DAYS.map((day) => (
					<TouchableRipple
						key={day.key}
						onPress={() => handleDayPress(day.key)}
						style={styles.ripple}
						borderless
						disabled={disabled}>
						<Avatar.Text
							testID={
								selectedDays.includes(day.key)
									? 'activeDay'
									: 'inactiveDay'
							}
							size={36}
							color={
								disabled
									? theme.colors.onSurfaceDisabled
									: selectedDays.includes(day.key)
									? theme.colors.onPrimary
									: theme.colors.primary
							}
							style={
								disabled && selectedDays.includes(day.key)
									? styles.disabledActive
									: disabled &&
									  !selectedDays.includes(day.key)
									? styles.disabled
									: selectedDays.includes(day.key)
									? styles.active
									: styles.inactive
							}
							label={t(`form.daysOfTheWeekValues.${day.label}`)}
						/>
					</TouchableRipple>
				))}
			</View>
		</>
	)
}

const createStyles = (theme: MD3Theme) =>
	StyleSheet.create({
		container: {
			flexDirection: 'row',
			justifyContent: 'center',
			gap: 10
		},
		active: {
			backgroundColor: theme.colors.primary,
			borderColor: theme.colors.primary,
			borderWidth: 1
		},
		inactive: {
			backgroundColor: theme.colors.onPrimary,
			borderColor: theme.colors.primary,
			borderWidth: 1
		},
		ripple: {
			borderRadius: 50
		},
		disabledActive: {
			backgroundColor: theme.colors.onSurfaceDisabled,
			borderColor: 'transparent',
			borderWidth: 1
		},
		disabled: {
			color: theme.colors.onSurfaceDisabled,
			backgroundColor: 'transparent',
			borderColor: 'transparent',
			borderWidth: 1
		}
	})

export default DayOfTheWeekInput
