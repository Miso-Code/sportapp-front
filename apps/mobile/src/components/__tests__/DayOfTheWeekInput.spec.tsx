import React from 'react'
import renderer, { ReactTestRenderer, act } from 'react-test-renderer'

import DayOfTheWeekInput from '../DayOfTheWeekInput'
import { Avatar, TouchableRipple } from 'react-native-paper'

import { useTranslation } from 'react-i18next'

describe('DayOfTheWeekInput', () => {
	let component: ReactTestRenderer
	let onChange: jest.Mock

	beforeEach(() => {
		jest.clearAllMocks()
		onChange = jest.fn()
		component = renderer.create(
			<DayOfTheWeekInput
				value={['sunday', 'monday']}
				onChange={onChange}
				label='Test'
			/>
		)
	})

	afterEach(() => {
		component.unmount()
	})

	it('should render correctly', () => {
		expect(component.toJSON()).toMatchSnapshot()
	})

	it('should render active days', () => {
		const days = component.root.findAllByType(Avatar.Text)
		expect(days).toHaveLength(7)
		for (let i = 0; i < 2; i++) {
			expect(days[i].props.testID).toBe('activeDay')
		}
		for (let i = 2; i < 7; i++) {
			expect(days[i].props.testID).toBe('inactiveDay')
		}
	})

	it('should call onChange when a day is pressed', () => {
		const dayContainers = component.root.findAllByType(TouchableRipple)
		act(() => {
			dayContainers[0].props.onPress()
		})
		expect(onChange).toHaveBeenCalledWith(['monday'])
	})

	it('should deactivate a day when it is pressed', () => {
		const dayContainers = component.root.findAllByType(TouchableRipple)
		act(() => {
			dayContainers[0].props.onPress()
		})
		const days = component.root.findAllByType(Avatar.Text)
		expect(days[0].props.testID).toBe('inactiveDay')
	})

	it('should activate a day when it is pressed', () => {
		const dayContainers = component.root.findAllByType(TouchableRipple)
		act(() => {
			dayContainers[2].props.onPress()
		})
		const days = component.root.findAllByType(Avatar.Text)
		expect(days[2].props.testID).toBe('activeDay')
	})

	it('should render monday first if locale starts with monday', () => {
		useTranslation().i18n.changeLanguage('es-CO')
		act(() => {
			component.update(
				<DayOfTheWeekInput
					value={['sunday', 'monday']}
					onChange={onChange}
					label='Test'
				/>
			)
		})
		const days = component.root.findAllByType(Avatar.Text)
		expect(days[0].props.label).toBe('form.daysOfTheWeekValues.MONDAY')
		expect(days[0].props.testID).toBe('activeDay')
		expect(days[days.length - 1].props.testID).toBe('activeDay')
	})

	it('should disable all days if is disabled', () => {
		act(() => {
			component.update(
				<DayOfTheWeekInput
					value={['sunday', 'monday']}
					onChange={onChange}
					label='Test'
					disabled
				/>
			)
		})
		const dayContainers = component.root.findAllByType(TouchableRipple)
		for (let i = 0; i < 7; i++) {
			expect(dayContainers[i].props.disabled).toBe(true)
		}
	})
})
