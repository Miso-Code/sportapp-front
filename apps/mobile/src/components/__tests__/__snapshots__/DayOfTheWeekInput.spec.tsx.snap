// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayOfTheWeekInput should render correctly 1`] = `
[
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "rgba(28, 27, 31, 1)",
          "writingDirection": "ltr",
        },
        [
          {
            "fontFamily": "System",
            "fontSize": 14,
            "fontWeight": "400",
            "letterSpacing": 0.25,
            "lineHeight": 20,
          },
          false,
        ],
      ]
    }
  >
    Test
  </Text>,
  <View
    style={
      {
        "flexDirection": "row",
        "gap": 10,
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 50,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(103, 80, 164, 1)",
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            {
              "alignItems": "center",
              "justifyContent": "center",
            },
            {
              "borderColor": "rgba(103, 80, 164, 1)",
              "borderWidth": 1,
            },
          ]
        }
        testID="activeDay"
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                {
                  "color": "rgba(255, 255, 255, 1)",
                  "fontSize": 18,
                  "lineHeight": 18,
                },
                undefined,
              ],
            ]
          }
        >
          form.daysOfTheWeekValues.SUNDAY
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 50,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(103, 80, 164, 1)",
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            {
              "alignItems": "center",
              "justifyContent": "center",
            },
            {
              "borderColor": "rgba(103, 80, 164, 1)",
              "borderWidth": 1,
            },
          ]
        }
        testID="activeDay"
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                {
                  "color": "rgba(255, 255, 255, 1)",
                  "fontSize": 18,
                  "lineHeight": 18,
                },
                undefined,
              ],
            ]
          }
        >
          form.daysOfTheWeekValues.MONDAY
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 50,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 1)",
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            {
              "alignItems": "center",
              "justifyContent": "center",
            },
            {
              "borderColor": "rgba(103, 80, 164, 1)",
              "borderWidth": 1,
            },
          ]
        }
        testID="inactiveDay"
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontSize": 18,
                  "lineHeight": 18,
                },
                undefined,
              ],
            ]
          }
        >
          form.daysOfTheWeekValues.TUESDAY
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 50,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 1)",
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            {
              "alignItems": "center",
              "justifyContent": "center",
            },
            {
              "borderColor": "rgba(103, 80, 164, 1)",
              "borderWidth": 1,
            },
          ]
        }
        testID="inactiveDay"
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontSize": 18,
                  "lineHeight": 18,
                },
                undefined,
              ],
            ]
          }
        >
          form.daysOfTheWeekValues.WEDNESDAY
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 50,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 1)",
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            {
              "alignItems": "center",
              "justifyContent": "center",
            },
            {
              "borderColor": "rgba(103, 80, 164, 1)",
              "borderWidth": 1,
            },
          ]
        }
        testID="inactiveDay"
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontSize": 18,
                  "lineHeight": 18,
                },
                undefined,
              ],
            ]
          }
        >
          form.daysOfTheWeekValues.THURSDAY
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 50,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 1)",
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            {
              "alignItems": "center",
              "justifyContent": "center",
            },
            {
              "borderColor": "rgba(103, 80, 164, 1)",
              "borderWidth": 1,
            },
          ]
        }
        testID="inactiveDay"
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontSize": 18,
                  "lineHeight": 18,
                },
                undefined,
              ],
            ]
          }
        >
          form.daysOfTheWeekValues.FRIDAY
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 50,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 1)",
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            {
              "alignItems": "center",
              "justifyContent": "center",
            },
            {
              "borderColor": "rgba(103, 80, 164, 1)",
              "borderWidth": 1,
            },
          ]
        }
        testID="inactiveDay"
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              [
                {
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontSize": 18,
                  "lineHeight": 18,
                },
                undefined,
              ],
            ]
          }
        >
          form.daysOfTheWeekValues.SATURDAY
        </Text>
      </View>
    </View>
  </View>,
]
`;
